Feature/65880 introduce portfolio proposal to project

ulferts opened the pull request:
# Ticket

https://community.openproject.org/wp/65880

ulferts committed:
add portfolio_proposal model

ulferts committed:
add portfolio column

ulferts committed:
add portfolio proposals to the side menu

ulferts committed:
fix routes on updating portfolio filters

EinLama commented on app/models/portfolio_proposal.rb:
```diff
@@ -0,0 +1,70 @@
+# ++
+
+class PortfolioProposal < ApplicationRecord
+  belongs_to :portfolio, -> { where(type: :portfolio) }, class_name: 'Project'
```
We have changed this column, it is now `project_type`:

```suggestion
  belongs_to :portfolio, -> { where(project_type: :portfolio) }, class_name: 'Project'
```

EinLama commented on app/models/portfolio_proposal_project.rb:
```diff
@@ -0,0 +1,48 @@
+
+class PortfolioProposalProject < ApplicationRecord
+  belongs_to :portfolio_proposal
+  belongs_to :project, -> { where(type: :project) }
```
We have changed this column, it is now `project_type`:

```suggestion
  belongs_to :project, -> { where(project_type: :project) }
```

EinLama commented on app/views/portfolio_managements/proposals/_form.html.erb:
```diff
@@ -0,0 +1,87 @@
+      <%= f.label :project_ids, t(:label_projects), class: 'form--label' %>
+      <div class="form--field-container">
+        <%= f.select :project_ids,
+                    Project.where(type: :project).map { |p| [p.name, p.id] },
```
We have changed this column, it is now `project_type`:

```suggestion
                    Project.where(project_type: :project).map { |p| [p.name, p.id] },
```

EinLama committed:
[#65880] type -> project_type

dombesz committed:
[#65879] Allow users to change order of projects in project list

https://community.openproject.org/work_packages/65879

ulferts committed:
start introducing portfolio specific table and row components to ease extending it

EinLama committed:
[#65895] add divider to project table action menu

EinLama committed:
[#65895] action menu stub for 'add to proposal'

ulferts committed:
fix highlighting proposal

EinLama committed:
[#65895] Show submenu for proposals

EinLama committed:
[#65895] Show project count in submenu

dombesz committed:
Manual sort order added with a lot of other useless things

EinLama committed:
[#65895] Link to proposal#new

EinLama committed:
[#65895] Add selected project to proposal

EinLama committed:
[#65895] Do not suggest proposals that already contain a project

EinLama committed:
Merge pull request #19580 from opf/feature/65895-add-project-to-portfolio-proposal-via-the-more-menu

Feature/65895 add project to portfolio proposal via the more menu

ulferts committed:
workflow based proposal handling

ulferts committed:
Merge remote-tracking branch 'origin/bmds_hackathon' into feature/65880-introduce-portfolio-proposal-to-project

dombesz committed:
Add a divider after the move actions items menus

EinLama committed:
[#65895] Adding to proposal without extra step

Known problem: flash messages will not show after successfully adding a
portfolio.

EinLama committed:
[#65895] Use correct State for Proposals

EinLama committed:
[#65895] Add all subprojects of a program or project to a proposal

EinLama committed:
Merge pull request #19594 from opf/feature/65895-streamline-portfolio-context-menu-process

[#65895] Adding to proposal without extra step

ulferts committed:
model complete process for portfolio proposal

toy committed:
hack links to stay in portfolio management

EinLama committed:
[#65880] Fix: remove duplicate lines added by git

ulferts committed:
have up/down only on proposals

EinLama committed:
[#65880] [#65895] Allow adding missing subprojects of a program

If a proposal already has SOME of the selected projects, we will now
offer the proposal anyways - only the missing projects will be added.

toy committed:
fix choosing roots to be scope based for ordering alongside hierarchy

EinLama committed:
Remove button within portfolios

dombesz committed:
Display rank and manual ordering by rank.

EinLama committed:
Redirect to the proposal after deletion

EinLama committed:
Redirect to portfolio page after context menu action

EinLama committed:
[#65880] Ensure only projects can be removed from a Proposal

dombesz committed:
Display rank column in the table and maintain rank order when manually ordering the table

EinLama committed:
[#65880] Create new proposals via dialog

toy committed:
make proposal an accessor in orders to not require changing every overrident order method

toy committed:
hide divider when not needed

dombesz committed:
Remove unnecessary rank column

ulferts committed:
remove now duplicate button from proposal header

toy committed:
fix hierarchy order url to not force projects path

ulferts committed:
include program itself when adding it to a proposal

dombesz committed:
Add acts as list to the rank column

toy committed:
don't blank ordering with hierarchy order link

ulferts committed:
adapt some labels

toy committed:
pass proposal in move action to table component

dombesz committed:
Add the proposal on project controller query when its present

ulferts committed:
add proposal first so that it gets the lowest rank

dombesz committed:
Add moving up and down while manually sorting by rank.

ulferts committed:
fix passing the proposal to the manual sorting order

ulferts committed:
rename manual sorting

dombesz committed:
Fix dynamic sort url generation

When a project is moved in rank or hierarchy, the rendered sort urls
were pointing to the move action instead of the index action. Hence the
hardcoding.

toy committed:
ensure correct urls?

ulferts committed:
sort by rank by default
